### DISCUSSION TITLE
How to Disable Embedding and Ensure Full Text Return from Files in Chat Using Tools?

### DISCUSSION THREAD

**Discussion by gmrx** [open-webui/open-webui#4655] [Q&A]
I'm currently facing an issue where I need to use the Tools class to translate the entire text from a file by sending it to chat. I'm using the code below, but the text is returned as a list. How can I disable embedding to ensure that the entire text from a file is returned as a single string? Here's the code I'm working with:

```python
import os
import requests
from datetime import datetime
from typing import List
from docx import Document
from pypdf import PdfReader
from pptx import Presentation
from unstructured.partition.auto import partition
import pypandoc

class Tools:
    def __init__(self):
        # If set to true it will prevent default RAG pipeline
        self.file_handler = True
        self.citation = True
        pass

    def extract_txt_content(self, file_path: str) -> str:
        try:
            with open(file_path, "r", encoding="utf-8") as file:
                return file.read()
        except Exception as e:
            return f"Error read: {e}"

    def get_files(self, __files__: List[dict] = []) -> str:
        result = ""
        for file_info in __files__:
            file_path = file_info["file"]["meta"]["path"]
            file_name = file_info["file"]["meta"]["name"]
            content_type = file_info["file"]["meta"]["content_type"]

            if content_type == "text/plain":
                content = self.extract_txt_content(file_path)
            else:
                content = (
                    f"Error type: {file_name} "
                )

            result += f"=== START FILE: {file_name} ===\n{content}\n=== END FILE: {file_name} ===\n\n"

        if isinstance(result, list):
            result = "\n".join(result)

        return result
```

How can I modify the implementation to ensure that the text is returned as a single complete string rather than a list? What adjustments should be made to prevent any unwanted processing like embedding that might be affecting the output?

**Comment by NeverOccurs** [comment]
Try this:
https://github.com/NeverOccurs/full_doc_pipe
Use it as a filter function in Open-Webui. It allows uploading a file and push it fully into system prompt. Then you can direct ask the LLM to summarize it or whatever. It doesn't support changing file midway through the chat though.
Still optimizing it but for now I think it is enough for my use case. I'm trying to incorporating it with Claude's prompt caching to minimize cost.
